062-topic-replication-factor-change.md

```diff
@@ -0,0 +1,223 @@
+From the point of view of the Topic Operator, a topic replication factor change can be in the following states:
+
+- Pending: `.spec.replicas` is updated, but the configuration change request is not sent to Cruise Control (`generation` > `observedGeneration`).
+- Ongoing: `.spec.replicas` is updated and Cruise Control has accepted the request, but the task is not completed (`generation` == `observedGeneration`).
+- Failed: `.spec.replicas` is updated with an invalid value, or there is a hard goal violation, or Cruise Control is not available (`generation` == `observedGeneration`).
+
+To request a replication factor change, the Topic Operator needs to authenticate and send HTTP requests to the Cruise Control's REST API.
+For that reason, we introduce the following environment variables, which will be set automatically by the Cluster Operator, or manually by the user in case of standalone deployment.
+
+| Name                                   | Type    | Default | Description                                           |
+|:---------------------------------------|---------|---------|-------------------------------------------------------|
+| STRIMZI_CRUISE_CONTROL_ENABLED         | Boolean | false   | Whether Cruise Control integration is enabled         |
+| STRIMZI_CRUISE_CONTROL_NAMESPACE       | String  | ""      | Namespace where Cruise Control is running             |
+| STRIMZI_CRUISE_CONTROL_HOSTNAME        | String  | ""      | Cruise Control hostname                               |
+| STRIMZI_CRUISE_CONTROL_PORT            | Integer | 9090    | Cruise Control port                                   |
+| STRIMZI_CRUISE_CONTROL_SSL_ENABLED     | Boolean | false   | Whether Cruise Control SSL encryption is enabled      |
+| STRIMZI_CRUISE_CONTROL_AUTH_ENABLED    | Boolean | false   | Whether Cruise Control Basic authorization is enabled |
+| STRIMZI_CRUISE_CONTROL_SECRET_NAME     | String  | ""      | Secret name containing Cruise Control truststore      |
+| STRIMZI_CRUISE_CONTROL_API_SECRET_NAME | String  | ""      | Secret name containing Cruise Control API credentials |
+| STRIMZI_CRUISE_CONTROL_RACK_ENABLED    | Boolean | false   | Whether the target Kafka cluster has rack awareness   |
+
+The topic configuration endpoint is asynchronous, which means that it validates the task, but the execution will be queued.
+Inside Cruise Control, a valid topic configuration task is first create in an `Active` state, then moved to `InExecution`, and finally to `Completed` or `CompletedWithError`.
+Only a single task can be `InExecution` state at any given time, which includes both topic configuration and rebalance tasks.
+
+When Cruise Control integration is enabled, a `ReplicationFactorManager` is created to manage the communication with the REST API.
+The `ReplicationFactorManager` stores all ongoing replication factor changes in an in-memory cache, in order to periodically check their status.
```

Comment:
How do you recover it after restart? If you can recover it after restart, is _in-memory_ cache really something you need?

Reply:
Yes, I need the pending changes cache to coordinate status updates between `BatchingTopicController` and the new `ReplicasChangeManager`. The ongoing changes cache is used to periodically check task status. Proposal updated.

Reply:
But why do you need the `ReplicasChangeManager`? Why can't you drive it from the `BatchingTopicController` directly? 
* You get a topic event (either because it changed or periodical)
* If it contains a RF change, you check if you are already running some proposal for it, what its state is and whether it is complete and if not, you either start the proposal or keep it for the next time.

Reply:
I need it to isolate the CC integration logic. Unlike with other topic configurations, the RF change reconciliation is asynchronous.

Reply:
I'm not sure I understand why that matters. You need to be able to recover the state anyway -> either from the CRs or from CC. So when you can recover the state, you can do it within the regular reconciliation procedure.